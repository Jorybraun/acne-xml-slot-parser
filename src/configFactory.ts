const fs = require('fs');
const jsonfile = require('jsonfile');

interface KeyPair {
    [key: string]: string;
}

interface ConfigOutput {
    keepPath: string;
    removePath: string;
}

interface ConfigInterface {
    toKeep: KeyPair[];
    toRemove: KeyPair[];
    keepAlive: boolean;
    inputPath: string;
    outputPath: ConfigOutput
}

interface ConfigConstructor {
    new  (
        toRemove: KeyPair[],
        toKeep: KeyPair[],
        keepAlive: boolean,
        inputPath: string,
        outputPath: ConfigOutput
    ): ConfigInterface
}

export const configFactory = (
    ctor: ConfigConstructor,
    toKeep: KeyPair[],
    toRemove: KeyPair[],
    keepAlive: boolean,
    inputPath: string,
    outputPath: ConfigOutput
) => {
    return new ctor(toRemove, toKeep, keepAlive, inputPath, outputPath)
};

export class Config implements ConfigInterface {
    constructor(
        public toKeep: KeyPair[],
        public toRemove: KeyPair[],
        public keepAlive: boolean,
        public inputPath: string,
        public outputPath: ConfigOutput
    ) {}
}

export const createConfig = (
    toKeep: KeyPair[] = [{ 'key': 'this is an example' }],
    toRemove: KeyPair[] = [{ 'key': 'this is an example' }],
    keepAlive: boolean = false,
    inputPath: string = '.please input a input path here',
    outputPath: ConfigOutput = {
        keepPath: '.please input a path here',
        removePath: '.please input a path here'
    }
): Promise<ConfigInterface> => {
    const config = configFactory(Config, toKeep, toRemove, keepAlive, inputPath, outputPath)
    return jsonfile.writeFile('./config.json', config, { spaces: 2 });
}