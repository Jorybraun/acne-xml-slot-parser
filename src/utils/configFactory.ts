/// <reference path="../types/jsonfile/index.d.ts" />
import { ConfigConstructor, ConfigInterface, ConfigOutput, KeyPair } from '../interfaces/Config';
// @ts-ignore
import * as jsonfile from 'jsonfile';

export const configFactory = (
    ctor: ConfigConstructor,
    { toKeep, toRemove, siteSpecificKey, keepAlive, inputPath, outputPath }: ConfigInterface
) => {
    return new ctor(toKeep, toRemove, siteSpecificKey, keepAlive, inputPath, outputPath);
};

export class Config implements ConfigInterface {
    constructor(
        public toKeep: KeyPair[],
        public toRemove: KeyPair[],
        public siteSpecificKey: string,
        public keepAlive: boolean,
        public inputPath: string,
        public outputPath: ConfigOutput
    ) {}
}

export const createConfig = (configInput: ConfigInterface, configOutputPath: string): Promise<void> => {
    const configOutput = configFactory(Config, configInput);
    return  jsonfile.writeFile(configOutputPath, configOutput, { spaces: 2 });
}

