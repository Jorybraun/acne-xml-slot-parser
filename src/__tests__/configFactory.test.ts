import { createConfig, configFactory, Config } from "../configFactory";
import * as fs from 'fs';

const testConfig = {
    toKeep: [ { bob: 'archer' } ],
    toRemove: [ { bob: 'archer' } ],
    keepAlive: true,
    inputPath: './file.xml',
    outputPath: { keepPath: './file', removePath: './remove' }
}

describe('configFactory', () => {

    it('should take in paramaters and return a config', () => {
        const config = configFactory(
            Config,
            [{'bob': 'archer'}],
            [{'bob': 'archer'}],
            true,
            './file.xml',
            {
                keepPath: './file',
                removePath: './remove'
            }
        );
        expect(config).toMatchSnapshot()
    });

    it('should create a config file', () => {
        createConfig(
            [{'bob': 'archer'}],
            [{'bob': 'archer'}],
            true,
            './file.xml',
            {
                keepPath: './file',
                removePath: './remove'
            }
        )
        .then(() => {
            const config = JSON.parse(fs.readFileSync('./config.json', 'utf-8'));
            expect(config).toMatchSnapshot();
            fs.unlinkSync('./config.json');
        })
    });
})