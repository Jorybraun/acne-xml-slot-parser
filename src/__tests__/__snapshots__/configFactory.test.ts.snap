// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configFactory should take in paramaters and return a config 1`] = `
Config {
  "inputPath": "./file.xml",
  "keepAlive": true,
  "outputPath": Object {
    "keepPath": "./file",
    "removePath": "./remove",
  },
  "toKeep": Array [
    Object {
      "bob": "archer",
    },
  ],
  "toRemove": Array [
    Object {
      "bob": "archer",
    },
  ],
}
`;
